File: GitToolBox/src/main/java/zielu/gittoolbox/metrics/Jmx.java
package zielu.gittoolbox.metrics;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.jmx.JmxReporter;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import java.security.Permission;
import javax.management.MBeanTrustPermission;
import org.jetbrains.annotations.NotNull;

class Jmx {
  private static final String DOMAIN = "zielu.gittoolbox";
  private static final Logger LOG = Logger.getInstance(Jmx.class);

  private Jmx() {
    throw new IllegalStateException();
  }

  static void reporter(MetricRegistry registry) {
    if (shouldExport()) {
      JmxReporter reporter = JmxReporter.forRegistry(registry).inDomain(DOMAIN).build();
      reporter.start();
    }
  }

  static MetricsReporter reporter(@NotNull Project project, MetricRegistry registry) {
    if (shouldExport()) {
      String projectName = project.getName().replaceAll("\\W", "");
      JmxReporter reporter = JmxReporter.forRegistry(registry).inDomain(DOMAIN + "." + projectName).build();
      reporter.start();
      return new JmxMetricsReporter(reporter);
    } else {
      return MetricsReporter.EMPTY;
    }
  }

  private static boolean shouldExport() {
    SecurityManager securityManager = System.getSecurityManager();
    if (securityManager != null) {
      Permission permission = new MBeanTrustPermission("register");
      try {
        securityManager.checkPermission(permission);
        return true;
      } catch (SecurityException e) {
        LOG.warn("Cannot export JMX metrics", e);
        return false;
      }
    }
    return true;
  }

  private static class JmxMetricsReporter implements MetricsReporter {
    private final JmxReporter reporter;

    private JmxMetricsReporter(JmxReporter reporter) {
      this.reporter = reporter;
    }

    @Override
    public void dispose() {
      reporter.close();
    }
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/metrics/Jmx.kt
package zielu.gittoolbox.metrics

import com.codahale.metrics.MetricRegistry
import com.codahale.metrics.jmx.JmxReporter
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import java.security.Permission
import javax.management.MBeanTrustPermission

internal object Jmx {
  private const val DOMAIN = "zielu.gittoolbox"
  private val log = Logger.getInstance(Jmx::class.java)

  @JvmStatic
  fun startReporting(registry: MetricRegistry): MetricsReporter {
    return if (shouldExport()) {
      val reporter = JmxReporter.forRegistry(registry).inDomain(DOMAIN).build()
      reporter.start()
      JmxMetricsReporter(reporter)
    } else {
      MetricsReporter.EMPTY
    }
  }

  @JvmStatic
  fun startReporting(project: Project, registry: MetricRegistry): MetricsReporter {
    return if (shouldExport()) {
      val projectName = project.name.replace("\\W".toRegex(), "")
      val reporter = JmxReporter.forRegistry(registry).inDomain("$DOMAIN.$projectName").build()
      reporter.start()
      JmxMetricsReporter(reporter)
    } else {
      MetricsReporter.EMPTY
    }
  }

  private fun shouldExport(): Boolean {
    return System.getSecurityManager()?.let { securityManager ->
      val permission: Permission = MBeanTrustPermission("register")
      return try {
        securityManager.checkPermission(permission)
        true
      } catch (e: SecurityException) {
        log.warn("Cannot export JMX metrics", e)
        false
      }
    } ?: true
  }
}

private class JmxMetricsReporter(private val reporter: JmxReporter) : MetricsReporter {
  private val log = Logger.getInstance(JmxMetricsReporter::class.java)

  override fun dispose() {
    reporter.close()
    log.info("Disposed")
  }
}
